Fix rendering issues when built against the 10.14 SDK.

s4y opened the pull request:
Before this commit, the Core Text renderer relied on a legacy behavior
of NSView to keep its content across draws. setNeedsDisplayInRect: could
be used to draw over parts of the existing content as drawing commands
were received without needing to redraw old content.

Layer-backed views lose this behavior and may be asked to redraw any or
all of their content at any time, and layer backing becomes the default
for apps built against the macOS 10.14 SDK.

This change adds a way to draw to an intermediate NSImage and then
create a layer contents from that image. It's similar to the CGLayer
path, but I wasn't able to get the CGLayer path to work without hanging
or crashing when I scrolled. My best guess from looking at stack traces
is that using CGContextDrawLayerInRect to draw a layer into itself
doesn't actually copy pixels, but adds the self-draw as an action to be
performed when the CGLayer is drawn into a bitmap context. After a bunch
of scrolling, these actions stack deeper and deeper and either hang or
overflow the stack when drawn.

When on, the new code skips -drawRect: by returning YES from
-wantsUpdateLayer. -updateLayer draws into an NSImage and then creates a
layer contents from the NSImage. On my machine, scrolling a large
document seems to be faster and use less CPU vs drawRect, which is
promising.

The new code is controlled by the MMUseCALayer user default, which is on
by default in this change. Fixes #751.

chdiza commented:
I just tried building with this on High Sierra with Xcode 9.4.1 just to make sure it doesn't cause any regressions.  It built, but as soon as macvim is launched (via `mvim`), I see the following at my terminal prompt:
```
2018-10-09 09:37:14.141 MacVim[15908:90254] Unable to find feedback application.
2018-10-09 09:37:14.322 MacVim[15908:90254] Unable to find feedback application.
```

chdiza commented:
Well, now those messages are still there even after I rebuilt without the patch.  They persist after a restart.  They were not present until I built with the patch.  I.e., building 8.1.451 on High Sierra with Xcode 9.4.1 worked fine.  Then I did it against the patch of this PR and the warnings started.  Then I rebuilt without the patch and they're still there.

s4y commented:
@chdiza Thanks for testing it out. I’m not sure what to say if the messages are still there after removing my patch. I searched for that message online and didn’t find anything conclusive.

s4y commented:
@chdiza how does everything look on your system aside from that message?

chdiza commented:
> how does everything look on your system aside from that message?

I don't know, I reverted the instant I saw that message.  I'm  a little worried to try it on my other High Sierra machine that has Xcode 10 instead of 9.4.1, in case the messages start appearing there too.

s4y commented:
Hmm, the patch is pretty minimal. Maybe someone else is willing to give it a shot? I have VMs but I’m not currently near my computer.

chdiza commented:
Eventually the messages went away.  So, thinking it was some fluke that just happened to coincide with making a test build, I tried building with the patch again.  And the messages immediately are back.

chdiza commented:
But just for fun I tried to see if the patch messes up non-native fullscreen mode.  And it does, tremendously.  Here are two screenshots, the first is not in fullscreen and the second is with the same file put into non-native fullscreen.

![reg](https://user-images.githubusercontent.com/3311113/46682587-ef5d6b80-cbbb-11e8-9a15-b4726fff6e95.png)

![screen shot 2018-10-09 at 12 05 51 p](https://user-images.githubusercontent.com/3311113/46682597-f84e3d00-cbbb-11e8-96e0-53158cdebd08.png)

s4y commented:
Yikes. I’ll take a look at fullscreen when I’m back at a computer. Still no idea about the messages.

chdiza commented:
I'll test this out later, once I'm at the machine that spews messages.  Thanks!

>  I think that this is in a land-able state if the occasional enter/exit flicker is the only remaining problem

Do we have any actual reports of this PR working on machines other than @s4y's?  Is it causing "messages" on anyone else's machine?

mario-grgic commented:
I have just tried this on macOS Mojave, compiled with Xcode 10, and it works fine. No error messages printed when running command line Vim either.

It fixes the Core Text renderer issue. No flashing in the GUI.

chdiza commented:
OK, I just built and *very* briefly tried it out on High Sierra w/ Xcode 941.

Re the flashing: on my older, slower machine I can see that what's happening during the flash  (while *entering* non-native FS) is that first the MacVim window expands and looks exactly like the second of my two screenshots, where everything is "stretched" out of proportion; then it quickly "fixes" itself and the fonts and character cells are the correct size, as are the &rows and &columns.

(Maybe that is a clue that will help in diagnosing the flashing.)

I hope to also test this out on Mojave later at a different machine.

amadeus commented:
Just as an FYI - performance on a 5k retina display (using this branch), when the window is maximized or full screen is actually way slower than the version built with the older SDK. Here's a video example. I tab between both version - the one with the white macOS bar at the top is the current version from the releases section (built using the old SDK) and the one with the dark theme title bar is from this branch:

http://wtf.amadeus.wtf/monosnap/screencast_2018-10-10_22-00-15.mp4

Just as an FYI - I am just holding down `j` and `k` for moving the cursor up and down - not tapping it. and `<c-y>` and `<c-e>` for scrolling.

As you can see, the version not from this branch is significantly faster and more performant.

s4y commented:
@amadeus I just put up a new version of this change which is somewhere in between the old code and the first version of my change (i.e. still uses `drawRect:` to do partial updates). It should perform better — would you mind giving it a try?

@chdiza Mind giving this new version a try, too?

s4y commented:
Small change: since this PR no longer does anything layer-specific, I just renamed the preference key and removed `self.wantsLayer = YES`. Shouldn't change behavior.

mario-grgic commented:
Ok, tried the latest changes, things work, no flashing. Performance on 2015 5k iMac is fine here. However, transparency is now broken. Window is not transparent with ```set transparency=50```.

amadeus commented:
@s4y unfortunately performance still appears to be pretty slow - here's another video example comparing the versions:  (dark theme header is built off commit `0015e854b`)

[video](http://wtf.amadeus.wtf/monosnap/screencast_2018-10-14_03-16-29.mp4)

amadeus commented:
If I build off of master, rendering is completely screwed up. However it's super responsive as you can see here:

[video](http://wtf.amadeus.wtf/monosnap/screencast_2018-10-14_03-31-58.mp4)

As a follow up, are you ensuring you only have 1 document open and the window is either full screened or resized to the full width of the screen?

mario-grgic commented:
@amadeus Ok, I see the same behavior in native full screen as well. Scrolling speed is much slower, and just moving cursor up down is much slower. In Windowed (not full screen), the scrolling speed is not that bad.

chdiza commented:
> Mind giving this new version a try, too?

I will do this when I'm back at the relevant machine this week :)

chdiza commented:
I just tested the latest iteration on High Sierra with Xcode 941 (non-retina screen).  This version is by far the worst; it's unusable.  The whole MacVim window often blacks out (not fixable with `:redraw`), and when it isn't blacked out, the antialiasing of the text is worse than what it would be on Mojave with a non-retina screen.

chdiza commented:
OK, I've also just tested it on Mojave (non-Retina).  It blacks out the window frequently, as described above when I tried it on High Sierra.  No `:redraw` or Ctrl-L will fix it.

But also, it "undoes" the fix for how bad fonts look on non-Retina screens on Mojave.  [This trick](https://ahmadawais.com/fix-macos-mojave-font-rendering-issue/) causes some, but not all, apps to make fonts look pretty much like they used to look pre-Mojave.  Much to my relief, the trick works on the master branch of MacVim, but is apparently somehow blocked by the current iteration of this PR.

Moreover, the present state of the PR somehow drastically worsens the appearance of fonts in MacVim even on High Sierra (non-Retina).  It makes them look like they do on Mojave non-Retina, i.e., thin, faint, and blurry.  This was unexpected.

s4y commented:
@chdiza That’s not how the PR is supposed to look/work and doesn’t match what I see. It might be an SDK or OS compatibility issue, but I have to investigate. Thanks for trying it.

@amadeus Thanks, that’s also very good to know.

Based on profiling, the pure `drawRect:` path is *way* faster than drawing to an image: the APIs I’m using in the current PR seem to make a copy of the image every time it’s drawn to, which makes it entirely unacceptable 😶.

And `scrollRect:by:` does some trickery that, as far as I know, won’t work in a layer-backed world (or, at least, I don’t know how to replicate it).

If we can find a way to draw to a thing without copies, and then display that in `drawRect:` or similar, or use it as a layer contents such that we can invalidate specific rects, that might be good enough.

Another option is to store state so that the view can draw whatever rects are requested at any time.

I’m not sure how much time I’ll have to work on this in the near future, but I will if I can. It involves corners of the SDK that I don’t know super well and takes some digging.

s4y commented:
Oh, one more thing: re. transparency, the technique of drawing transparent pixels in `drawRect:` to punch holes in the window *doesn’t work* in a layer-backed world, AFAIK. The window itself needs to be transparent. Thst’s another issue :/.

chdiza commented:
@s4y Thanks, by the way, for all your hard work on this :)

amadeus commented:
While I don't really have much say, I think it would be awesome to have someone like you who knows what they are doing, spend some time really creating a great renderer for MacVim! A top level LICENSE file should be super quick to add and I hope it does not become a blocker!

ychin commented:
Hi, sorry for the late comment. I just tried the patch out and I'm a little concerned about the decreased scrolling performance. I think as long as we are doing software-based blitting in `scrollRect` (which I think the current patch is doing) we would keep having the same issue. I guess as you said we could just cache the state of the text buffer and redraw, which should be faster than doing image copies on the CPU. From playing around with VimR I believe this is what they do as well, and VimR is fast-ish when using in fullscreen (not as fast as the old MacVim using `scrollRect:by:` in pre-Mojave though).

I'm curious about the comment by @chdiza about font rendering being broken (i.e. not doing subpixel font smoothing). When I played around with this before Mojave came out, it was quite hard to rendering text to an offscreen bitmap and tell macOS to do subpixel rendering (this is because the nature of subpixel rendering requires per-color blending which is hard to composite using standard RGBA buffers). I did manage to follow [this](https://stackoverflow.com/questions/4293487/how-to-work-around-poor-text-rendering-in-text-backed-by-a-calayer) by initializing an image with [kCGBitmapByteOrder32Host](https://developer.apple.com/documentation/coregraphics/kcgbitmapbyteorder32host) which seemed to be the only way to get macOS to respect subpixel font smoothing. If that doesn't work maybe this rendering mode should only be enabled for Mojave? It would still break the trick (hack?) to turn on subpixel font smoothing for Mojave but I suspect that toggle will be killed by Apple soon anyway since they seem to have decided to drop subpixel font smoothing as a feature :( .

I still think a permanent solution is we need more custom control for the rendering stack (e.g. a Metal renderer) so we can easily accelerate the scrolling (copying images in GPU is very fast) but that will take much more work obviously.

amadeus commented:
We have someone who is knowledgeable and willing to work on this serious issue here, and all that is required is a simple license file, I'd hate for that to continue to be a blocker here. It's nearly been 10 days and there's been nothing from any of the maintainers about this. Can we please get the ball moving on this? It's a pretty major problem.

I am totally willing to put together a license file provided I have some direction on the requirements, but this really shouldn't be hard to get done and it can enable s4y to start working on this. 😬

s4y commented:
@amadeus Thanks for the support 👍!

@ychin I haven’t had a chance to write a longer answer but, summary, I agree that the current version isn’t acceptable perf-wise and with pre-Mojave subpixel smoothing. I’m definitely willing to spend time exploring redrawing. I’m also interested in investigating `scrollRect:by:` to see if I can find a similarly-performant equivalent. I’m *also* interested in revisiting something I mentioned in another comment — using multiple layers to represent the content and moving them around. I *might* also be willing to explore GPU rendering, but can’t guarantee it. The extra background in your comment is super useful.

ychin commented:
@amadeus We will look into the LICENSE issue (#765).

@s4y We should just try to just merge this change in first. My comments were mostly just for discussion for future enhancement :). Slow rendering is still better than broken rendering…

I'm probably going to take over reviewing this change. I'm still looking through the whole change, but a couple feedbacks now:

1. Instead of a static `MMDrawToImage` that defaults to true, can you make it depend on whether Mojave is enabled? If MacVim is either compiled with pre-10.14 SDK (`MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_X_VERSION_10_14`) *or* at runtime it's used in a pre-Mojave OS (`floor(NSAppKitVersionNumber) <= NSAppKitVersionNumber10_13`) I think we should automatically use the old rendering code if the user didn't set `MMDrawToImage` manually. This minimizes the damage from the slower performance.
2. Did you have a chance to look into fixing subpixel rendering? I don't think it's a dealbreaker if we do (1) since subpixel rendering is not officially supported on Mojave anyway and the current way to turn it on it's really a hack. It would still be nice to find a way to fix it though so people who use the hidden toggle to turn it on can't still use it. On the other hand I feel that we definitely shouldn't break subpixel rendering for older OSes.
3. We could consider ditching the whole CGLayer code and just use the new draw to image code but gut feeling is it would cause some backwards compatibility issues, so maybe we should punt on that for now unless we can verify otherwise.

A side comment: I spent a lot time before Mojave came out trying to find a replacement for `scrollRect:by:` and I don't think there's a good alternative, since Apple really wants you to back everything by CALayer's and let the Cocoa handle scrolling for you instead of a custom rect to copy. I think your per-line layer scrolling could work, but let's do that later since that would complicate the logic by quite a bit.

ychin commented:
@s4y There is now a top-level LICENSE file (#765). Did you have time to look at my suggestions for change? I just want to make sure we don't regress on older OS versions.

s4y commented:
That's great, thanks so much. I'm planning to work on this around 1d/week, hopefully starting this Friday.

ychin commented:
Pinging @s4y. Do you mind pushing your WIP branches to this PR if you have time? I can fix up the remaining issues as I would like this fix this soon, but I don't want to take code from un-pushed changes for miscellaneous liability reasons.

s4y commented:
@ychin et. al: I think this is in a good state now, PTAL!

s4y commented:
(And, to all, feel free to grab this and try to poke holes in it.)

amadeus commented:
Just built it and ran it. So far the best version yet, however it still feels a bit slower than the previous version (the version built on the old SDK - downloaded from releases). And the wider the window the worse performance gets, even though the lines don't continue all the way offscreen.

I've included a few videos showing these differences (videos with the light title bar are the old SDK version, and the dark theme bar are this new build). I have a very fast keyboard repeat rate, and I use this to show the differences in performance, since this version performs a bit slower than the old version. Also worth noting all these were conducted on a late 2015 iMac Retina.

### Moving cursor up and down by holding `j`/`k`
[Old SDK](http://wtf.amadeus.wtf/monosnap/screencast_2018-12-02_11-48-17.mp4)
[New SDK](http://wtf.amadeus.wtf/monosnap/screencast_2018-12-02_11-48-53.mp4)

The window is full height and half the screen size during this test. You'll notice a small bit of jank in moving the cursor up and down in the new version, vs feeling much smoother on the old version.

### Scrolling document by holding `<c-y>`/`<c-e>`
[Old SDK](http://wtf.amadeus.wtf/monosnap/screencast_2018-12-02_11-50-19.mp4)
[New SDK](http://wtf.amadeus.wtf/monosnap/screencast_2018-12-02_11-52-27.mp4)

In this test, the window was maximized (but not full screen). The new version scrolls significantly slower than the old version. Performance gets marginally better if I make a ton a vertical splits and narrow the document. Not sure how feasible it is, but it would be awesome if we could overall improve performance from the old version, but that may be out of scope for this project.

s4y commented:
@amadeus Thanks much. I wanted to focus on finishing this version — see https://github.com/macvim-dev/macvim/pull/757#issuecomment-434961122. However, I have some work in progress that I wager will significantly improve performance from the old version. So, I'm glad this hasn't broken anything, and stay tuned 😉.

chdiza commented:
I just tried this.  Something is wrong with the cursor positioning after scrolling.  I open a document, and the cursor is on line 1.  I do <C-F> to go down to the end.  I do <C-b> to go all the way back up.  And then the cursor is on line 17 (out of 55 visible whole lines).  I feel fairly sure that the cursor should either end up on line 1 or on the last screen line.

OTOH maybe I am wrong to expect that.  I just tried plain vim/vim on the same file, and the cursor winds up on line 41.

So this seems *probably* unrelated to the PR.  But the discrepancy between vim and MacVim seems worth noting.

chdiza commented:
I'm also seeing some window blackouts when coming out of fullscreen (non-native mode).

ychin commented:
I can see some of the issues with blacking out and this change not working properly with the zoom button etc. I have a work-in-progress fix, and will push shortly, along with some other fixes for this PR.

@amadeus The performance fix will likely have to wait. It will require a bit of change.

@chdiza Can you elaborate more about the `Ctrl-F` / `Ctrl-B` issue? I couldn't repro it.

ychin commented:
I have pushed my changes that should fix the misc resizing issues related to this change. That means zoom button, drag-to-resize, changing font size (Cmd+=), fullscreen (non-native and native), etc should all "just works"™ now.

I believe @s4y also fixed this to work with subpixel font smoothing, but @chdiza feel free to confirm. :)

I feel pretty good about this patch, but as usual there's probably something that I missed, so I'll let it soak a little bit and see if others or @s4y have additional comments or find issues with the current patch before I merge.

Performance: It's not ideal, but I think it's within tolerable range. Let's just fix the brokeness first. It's been broken for Mojave Homebrew users for months which I will try to make sure not to happen again.

ronen commented:
> see if others or @s4y have additional comments

Comment from the peanut gallery:  I haven't done any extensive testing, but I've been using this all day and it's been working fine.   Thanks!!!

pablinos commented:
> Comment from the peanut gallery: I haven't done any extensive testing, but I've been using this all day and it's been working fine.

Same here - I noticed ctags being a bit more active (it's set to regenerate on save) than usual, but I don't think that's got anything to do with this change.

I can't say that I've noticed anything significant from a performance point of view too, but I haven't done any specific comparisons.

Thanks for working on it everyone!

ychin commented:
The change was merged! I'll go close the related issues and prep an update soon. Thanks @s4y for the help.
